File: src/Parser.cpp
#include "Parser.hpp"
#include <sstream>

Parser::Parser(Graph& g, Config& c) : graph(g), config(c) {
    library = config.getLibrary();
}

bool Parser::parse(const std::string& filePath) {
    std::ifstream file(filePath);
    if (!file.is_open()) {
        std::cerr << "Error: Could not open design file " << filePath << std::endl;
        return false;
    }

    std::string line;
    std::smatch match;

    // Regex patterns
    // 1. Input/Output ports: input [3:0] data_in; or input clk;
    std::regex portRegex(R"((input|output)\s+(?:\[\d+:\d+\]\s+)?(\w+)\s*;)");
    
    // 2. Instance: CellType instName (.Pin(Net), ...);
    // Simple regex to catch the line. 
    // Capturing explicit pin mapping requires parsing the content inside ().
    std::regex instRegex(R"(^\s*(\w+)\s+(\w+)\s*\((.*)\)\s*;)");

    // 3. Pin connection: .Pin(Net)
    std::regex pinRegex(R"(\.(\w+)\s*\(([^)]+)\))");

    // 4. Register declaration (ignored for graph, handled via instance logic) but we might need to know if something is a wire or reg?
    // In this structural netlist, mostly wires and instances.

    std::cout << "Parsing Verilog: " << filePath << std::endl;

    while (std::getline(file, line)) {
        // Strip comments // 
        size_t commentPos = line.find("//");
        if (commentPos != std::string::npos) {
            line = line.substr(0, commentPos);
        }

        if (std::regex_search(line, match, portRegex)) {
            std::string direction = match[1];
            std::string name = match[2];
            // Note: Vector ports like [3:0] data_in are treated as single node "data_in" in this simplified parser? 
            // OR we should expand them? 
            // In the Python version: "data_in[0]" is created if referenced.
            // But the port declaration "input [3:0] data_in" creates "data_in" ???
            // Let's check Python parser. It uses Pyverilog which expands things. 
            // For simplicity here, we will create nodes lazily when we see them in instances.
            // But for PRIMARY INPUTS, we need to know they drive nets.
            // Let's defer port creation to connection phase or create generic "port" type.
            // Actually, if we see `input clk`, we know `clk` is a net driver.
            
            // For this practice, we assume scalar or vector nets are handled by name string match.
            // e.g., "data_in[0]" is a distinct name from "data_in[1]".
            // If the line says "input [3:0] data_in", we technically should expand it. 
            // But simplistic regex won't expand. 
            // *Simplification*: We'll assume the netlist uses explicit indices in instances, e.g. .D(data_in[0]).
            // So we treat "data_in[0]" as a net name.
            // We need to mark primary inputs/outputs though.
            
            // If we match "input", we can't easily expand without parsing [3:0].
            // Let's ignore this for now and handle it via `processInstance` and lazy creation, 
            // EXCEPT we need to know what are Primary Inputs for Arrival Time.
            // We can infer Primary Inputs: nodes with 0 in-degree that are not constants? 
            // Or just rely on "input_delay" constraints applied to all inputs?
            
            // Hack for now: We won't strictly validate port directions vs declaraction.
            continue; 
        }

        if (std::regex_search(line, match, instRegex)) {
            std::string cellType = match[1];
            std::string instName = match[2];
            std::string content = match[3];
            
            if (cellType == "module") continue; // Skip module declaration

            std::map<std::string, std::string> connections;
            
            // Parse pin connections .Pin(Net)
            auto pins_begin = std::sregex_iterator(content.begin(), content.end(), pinRegex);
            auto pins_end = std::sregex_iterator();
            
            for (std::sregex_iterator i = pins_begin; i != pins_end; ++i) {
                std::smatch pinMatch = *i;
                connections[pinMatch[1]] = pinMatch[2];
            }
            
            processInstance(cellType, instName, connections);
        }
    }
    
    connectNets();

    return true;
}

void Parser::processInstance(const std::string& cellType, const std::string& instName, const std::map<std::string, std::string>& connections) {
    if (!library["cells"].contains(cellType)) {
        std::cerr << "Warning: Unknown cell type " << cellType << " for instance " << instName << std::endl;
        return;
    }

    // Explicitly convert json value to object (std::map-like access)
    json cellInfo = library["cells"][cellType];
    
    // Process pins
    for (auto const& [pin, net] : connections) {
        std::string pinName = instName + "/" + pin;
        Node* pinNode = graph.getOrCreateNode(pinName, "pin");

        bool isInput = false;
        bool isOutput = false;

        // Check inputs
        for (auto& inp : cellInfo["inputs"]) {
            if (inp.get<std::string>() == pin) isInput = true;
        }
         // Check outputs
        for (auto& outp : cellInfo["outputs"]) {
            if (outp.get<std::string>() == pin) isOutput = true;
        }

        if (isInput) {
            net_loads[net].push_back(pinNode);
        } else if (isOutput) {
            net_drivers[net].push_back(pinNode);
        }
    }

    // Internal Edges
    bool isSeq = false;
    if (cellInfo.contains("is_seq") && cellInfo["is_seq"].get<bool>()) isSeq = true;

    if (!isSeq) {
        double delay = cellInfo["delay"];
        for (auto& outPin : cellInfo["outputs"]) {
            std::string outPinName = instName + "/" + outPin.get<std::string>();
            Node* outNode = graph.getOrCreateNode(outPinName, "pin");

            for (auto& inPin : cellInfo["inputs"]) {
                std::string inPinName = instName + "/" + inPin.get<std::string>();
                Node* inNode = graph.getOrCreateNode(inPinName, "pin");
                
                inNode->addEdge(outNode, delay, "internal");
            }
        }
    }
}

void Parser::connectNets() {
    double fanoutFactor = 0.0;
    if (library.contains("wire_load_model")) {
        fanoutFactor = library["wire_load_model"]["fanout_factor"];
    }

    for (auto const& [net, drivers] : net_drivers) {
        if (net_loads.find(net) != net_loads.end()) {
            std::vector<Node*>& loads = net_loads[net];
            double delay = loads.size() * fanoutFactor;

            for (Node* driver : drivers) {
                for (Node* load : loads) {
                    driver->addEdge(load, delay, "net");
                }
            }
        } else {
            // Net has drivers but no loads (maybe output port?)
            // In this parser, output ports are not explicitly nodes unless we parse "output ..."
            // Ideally we need to create a Port node and connect driver to it.
            // Let's create a "port" node for the net itself if it looks like a port.
            // Or just leave it open. STA usually checks endpoints at DFF/Ports. 
            // If the net name is "sum_out", we want to see it as an endpoint.
            // Let's create a dummy load node for the net if it's potentially an output?
            // "sum_out" is driven by register Q.
            // Let's create a node for the Net itself? No, standard graph is Pin-to-Pin.
            
            // Special handling for output ports:
            // If a generic net has no loads, we treat the net name as a "Output Port Node" 
            // and connect driver to it.
            Node* portNode = graph.getOrCreateNode(net, "port");
            for (Node* driver : drivers) {
                 driver->addEdge(portNode, 0.0, "net"); // Connect driver to output port
            }
            // Mark as port for sure
            portNode->type = "port";
        }
    }
    
    // Handle Input Ports driving loads
    for (auto const& [net, loads] : net_loads) {
        if (net_drivers.find(net) == net_drivers.end()) {
            // Net has loads but no internal drivers -> Driven by Input Port
            Node* portNode = graph.getOrCreateNode(net, "port");
            double delay = loads.size() * fanoutFactor;
            
            for (Node* load : loads) {
                portNode->addEdge(load, delay, "net");
            }
        }
    }
}

File: sta_engine/parser.py
from pyverilog.vparser.parser import parse
from pyverilog.vparser.ast import InstanceList, Description, ModuleDef, Port, Ioport, Input, Output, Pointer, Identifier
import os
from typing import Dict, List, Any, Optional
from .graph import Graph, Node

class VerilogParser:
    """Parses Verilog designs and builds the STA Graph."""
    
    def __init__(self, library_config: Dict[str, Any]):
        self.lib = library_config
        self.graph = Graph()
        self.net_drivers: Dict[str, List[Node]] = {} # NetName -> [Node]
        self.net_loads: Dict[str, List[Node]] = {}   # NetName -> [Node]

    def parse(self, file_path: str) -> Graph:
        """Parses a Verilog file and returns the constructed STA Graph."""
        if not os.path.exists(file_path):
            raise FileNotFoundError(f"{file_path} not found")
            
        print(f"Parsing Verilog: {file_path}")
        try:
            ast, _ = parse([file_path])
            module_def: ModuleDef = ast.description.definitions[0]
            
            self._process_instances(module_def)
            self._process_ports(module_def)
            self._build_net_connections()
            
            return self.graph
        except Exception as e:
            raise RuntimeError(f"Failed to parse Verilog file: {e}") from e

    def _process_instances(self, module_def: ModuleDef):
        """Iterates over all instances in the module and processes them."""
        for item in module_def.items:
            if isinstance(item, InstanceList):
                for inst in item.instances:
                    self._process_single_instance(inst)

    def _process_single_instance(self, inst):
        """Processes a single instance to create nodes and internal edges."""
        cell_type = inst.module
        inst_name = inst.name
        
        if cell_type not in self.lib['cells']:
            print(f"Warning: Unknown cell type {cell_type} for instance {inst_name}")
            return

        cell_info = self.lib['cells'][cell_type]
        self._create_pin_nodes(inst, inst_name, cell_info)
        self._create_internal_timing_arcs(inst_name, cell_info)

    def _create_pin_nodes(self, inst, inst_name: str, cell_info: Dict[str, Any]):
        """Creates graph nodes for instance pins and registers net connections."""
        for port in inst.portlist:
            pin_name = f"{inst_name}/{port.portname}"
            net_name = self._resolve_net_name(port.argname)
            
            pin_node = self.graph.get_or_create_node(pin_name, "pin")

            is_input = port.portname in cell_info.get('inputs', [])
            is_output = port.portname in cell_info.get('outputs', [])

            if is_input:
                self.net_loads.setdefault(net_name, []).append(pin_node)
            elif is_output:
                self.net_drivers.setdefault(net_name, []).append(pin_node)

    def _create_internal_timing_arcs(self, inst_name: str, cell_info: Dict[str, Any]):
        """Creates internal edges based on cell timing arcs."""
        is_sequential = cell_info.get('is_seq', False)
        
        if not is_sequential:
            # Combinational logic: Create arcs from all inputs to all outputs
            delay = cell_info.get('delay', 0.0)
            for out_pin in cell_info.get('outputs', []):
                out_node = self.graph.get_or_create_node(f"{inst_name}/{out_pin}")
                for in_pin in cell_info.get('inputs', []):
                    in_node = self.graph.get_or_create_node(f"{inst_name}/{in_pin}")
                    in_node.add_edge(out_node, delay, "internal")
        else:
            # Sequential logic (DFF): No internal combinational arc from D to Q
            # Timing arcs like clk->Q are handled during analysis/arrival time propagation start points
            pass

    def _process_ports(self, module_def: ModuleDef):
        """Processes top-level module ports."""
        for port in module_def.portlist.ports:
            # Pyverilog AST navigation to find the actual port details
            first_level = port.first
            if isinstance(first_level, Ioport):
                first = first_level.first
                node = self.graph.get_or_create_node(first.name, "port")
                
                if isinstance(first, Input):
                     self.net_drivers.setdefault(first.name, []).append(node)
                elif isinstance(first, Output):
                     self.net_loads.setdefault(first.name, []).append(node)

    def _build_net_connections(self):
        """Creates edges between drivers and loads on the same net."""
        fanout_factor = self.lib.get('wire_load_model', {}).get('fanout_factor', 0.0)
        
        for net_name, drivers in self.net_drivers.items():
            if net_name in self.net_loads:
                loads = self.net_loads[net_name]
                fanout = len(loads)
                delay = fanout * fanout_factor
                
                for driver in drivers:
                    for load in loads:
                        driver.add_edge(load, delay, "net")

    def _resolve_net_name(self, argname: Any) -> str:
        """Resolves the net name from Pyverilog AST nodes."""
        if isinstance(argname, Pointer):
             return f"{argname.var}[{argname.ptr}]"
        elif isinstance(argname, Identifier):
             return f"{argname.name}"
        else:
            return str(argname)
